### ISSUE TITLE
python setup.py test is broken

### ISSUE THREAD

**Issue by cdeil** [iminuit/iminuit#213]
I see this error locally:
```
$ python setup.py test
running test
Traceback (most recent call last):
  File "setup.py", line 140, in <module>
    'build_ext': SmartBuildExt,
  File "/Users/deil/Library/Python/3.6/lib/python/site-packages/setuptools/__init__.py", line 129, in setup
    return distutils.core.setup(**attrs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/Users/deil/Library/Python/3.6/lib/python/site-packages/setuptools/command/test.py", line 216, in run
    cmd = ' '.join(self._argv)
TypeError: sequence item 2: expected str instance, bool found
```

I think this was there all the time, and we just didn't invocate the tests like that.

I think copy & pasting the latest code from here to update our setup.py would fix the issue:
https://docs.pytest.org/en/latest/goodpractices.html#manual-integration

Or we switch to this:
https://docs.pytest.org/en/latest/goodpractices.html#integrating-with-setuptools-python-setup-py-test-pytest-runner

I haven't looked at the pros and cons of the different ways to run tests in a while, so I'm not sure.
@HDembinski - Any opinions?
If not, we could keep as-is and I just do the update in `setup.py` to fix this.

**Comment by HDembinski** [comment]
@cdeil I looked into `setup.py` and apparently we went for 'manual-integration', it is just broken. I tried a few things to get it working again, and got some improvement, but then another failure. I don't really understand the error and how all this works. I am starting to find setuptools really annoying...

**Comment by HDembinski** [comment]
My preferred solution would be to drop setup.py and use cmake to build this project. This would solve several of the issues I have with the current build. I don't know how that would work with pypi integration?

**Comment by cdeil** [comment]
@HDembinski - The `setup.py` has to remain, it's what `pip` uses, the standard how Python packages are built. It is possible to call into other build systems from `setup.py`. I don't know if calling into `cmake` would mean an extra dependency on `cmake`, or if `cmake` generates something that just works without extra dependency on all platforms we target (like what Cython does). 

I think most or all people agree that setup.py isn't a great build system for C++.  But I don't think we should try to replace it, even if isn't possible somehow. Never touch a running system.

I'll try to find some time next week to look at how we run tests and find a solution to the issue mentioned above.

**Comment by HDembinski** [comment]
pybind11 has an example, where setup.py calls cmake to compile the code.
https://github.com/pybind/cmake_example/blob/master/setup.py
I think this is the way to go.
Yes, this would add an extra dependency on cmake, but cmake is a standard in the C++ world now, so it is something that people likely have installed already. It looks like you can even install cmake with PyPI
https://pypi.python.org/pypi/cmake

**Comment by HDembinski** [comment]
Sorry to disagree, but "never touch a running system" only make sense when
a) you cannot thoroughly test the system (but we can, we have CI for linux and windows now, and perhaps soon for OSX)
b) you don't understand the system

If you want to stick to setup.py, then let me know how we can set it up in such a way, that Minuit is build into a separate shared library that is linked to _libiminuit.so. I looked into the setuptools docu and could not find a way to compile a C++ shared libary and link another shared library to the first, all within setup.py.
If you can do all that with setuptools, ok, then I have no practical reason to move to cmake.

**Comment by cdeil** [comment]
I think a main reason why people like `iminuit` is that it has zero dependencies and is easy to install. IMO the concern is mainly distribution, it's not clear if install really still works smoothly if we need cmake. E.g. I have a Raspberry Pi, who knows is `pip install cmake` there works. And what if I did `sudo apt-get install cmake` there? -> an extra moving part is an extra moving part and some users will have problems.

Users don't care if Minuit2 is in a separate shared lib, no? The benefits I see of moving to cmake are 1. simplification of our setup.py and 2. faster build, because cmake uses multi-core. The drawback is the extra dependency / moving part. We can discuss the change if you want, in a new issue, but to me it doesn't seem like a good move.

**Comment by cdeil** [comment]
Whether the setuptools build can be improved: I don't know. There must be Python projects that have C++ that don't call out to cmake or other build systems. Maybe scikit-learn? This question could also be a good one for stackoverflow if it doesn't exist already.

**Comment by HDembinski** [comment]
Ok, you convinced me, no cmake then and we try to find out how to build two c++ libs that link to each other with cmake. And we still need to figure out how to fix this issue. My attempt failed and I would have to dig deeper. Perhaps someone with more experience in setuptools could give it a try?

**Comment by cdeil** [comment]
Fixed by @HDembinski  in #221.
